pkg/errors/errors.go

```diff
@@ -361,15 +361,75 @@ func ToServiceError(err error) *ServiceError {
 	}
 }
 
+type ServiceErrorBuilder interface {
+	Wrap(serviceError ServiceError) ServiceErrorBuilder
+	WithCode(code ServiceErrorCode) ServiceErrorBuilder
+	WithReason(reason string, values ...any) ServiceErrorBuilder
+	WithHttpCode(httpCode int) ServiceErrorBuilder
+	WithCause(cause error) ServiceErrorBuilder
+	Recoverable() ServiceErrorBuilder
+	Build() *ServiceError
+}
+
+var _ ServiceErrorBuilder = &serviceErrorBuilder{}
+
+func NewServiceErrorBuilder() ServiceErrorBuilder {
+	return &serviceErrorBuilder{}
+}
+
+type serviceErrorBuilder struct {
+	ServiceError
+}
+
+func (seb *serviceErrorBuilder) Wrap(serviceError ServiceError) ServiceErrorBuilder {
+	seb.WithReason(serviceError.Reason).
+		WithCode(serviceError.Code).
+		WithCause(&serviceError)
+	return seb
+}
+
+func (seb *serviceErrorBuilder) WithCode(code ServiceErrorCode) ServiceErrorBuilder {
+	seb.Code = code
+	return seb
+}
+
+func (seb *serviceErrorBuilder) WithReason(reason string, values ...any) ServiceErrorBuilder {
+	seb.Reason = fmt.Sprintf(reason, values...)
+	return seb
+}
+
+func (seb *serviceErrorBuilder) WithHttpCode(httpCode int) ServiceErrorBuilder {
+	seb.HttpCode = httpCode
+	return seb
+}
+
+func (seb *serviceErrorBuilder) WithCause(cause error) ServiceErrorBuilder {
+	seb.cause = cause
+	return seb
+}
+
+func (seb *serviceErrorBuilder) Recoverable() ServiceErrorBuilder {
+	seb.recoverable = true
+	return seb
+}
+
+func (seb *serviceErrorBuilder) Build() *ServiceError {
+	err := NewWithCause(seb.Code, seb.cause, seb.Error(), seb.Reason)
```

Comment:
Shouldn't the builder perform some verifications or at least set some defaults?

Reply:
The checks are done in the `NewWithCause` function.
The `recoverable` field defaults to `false`